File: __tests__/lib/goals.test.ts
// Tests de la logique pure des objectifs d'entraînement (lib/goals.ts) :
// bornes de période, calcul d'avancement, libellés, parsing robuste et
// opérations de liste. Pur (aucun accès base/React). Dates en heure locale
// pour rester indépendant du fuseau de la CI.
import {
  computeProgress,
  describeGoal,
  formatGoalValue,
  makeGoal,
  parseGoals,
  periodRange,
  removeGoal,
  serializeGoals,
  upsertGoal,
  type Goal,
} from '@/lib/goals';

const goal = (over: Partial<Goal> = {}): Goal => ({
  id: 'g1',
  metric: 'sessions',
  period: 'week',
  target: 3,
  activity: 'all',
  ...over,
});

describe('periodRange', () => {
  it('semaine : lundi 00:00 → lundi suivant (7 jours)', () => {
    const now = new Date(2025, 5, 11, 14, 30).getTime(); // mercredi 11 juin
    const { fromMs, toMs } = periodRange('week', now);
    const from = new Date(fromMs);
    expect(from.getDay()).toBe(1); // lundi
    expect(from.getDate()).toBe(9);
    expect(from.getHours()).toBe(0);
    expect(toMs - fromMs).toBe(7 * 86_400_000);
  });

  it('mois : 1er du mois → 1er du mois suivant', () => {
    const now = new Date(2025, 5, 11, 14, 30).getTime(); // juin
    const { fromMs, toMs } = periodRange('month', now);
    const from = new Date(fromMs);
    const to = new Date(toMs);
    expect(from.getDate()).toBe(1);
    expect(from.getMonth()).toBe(5); // juin
    expect(to.getDate()).toBe(1);
    expect(to.getMonth()).toBe(6); // juillet
  });

  it('mois : décembre déborde sur janvier de l’année suivante', () => {
    const now = new Date(2025, 11, 20).getTime(); // décembre 2025
    const { toMs } = periodRange('month', now);
    const to = new Date(toMs);
    expect(to.getFullYear()).toBe(2026);
    expect(to.getMonth()).toBe(0); // janvier
  });
});

describe('computeProgress', () => {
  it('avancement partiel : ratio borné, non atteint', () => {
    const p = computeProgress(goal({ target: 4 }), 1);
    expect(p.ratio).toBeCloseTo(0.25);
    expect(p.done).toBe(false);
    expect(p.value).toBe(1);
    expect(p.target).toBe(4);
  });

  it('atteint à exactement la cible', () => {
    const p = computeProgress(goal({ target: 3 }), 3);
    expect(p.done).toBe(true);
    expect(p.ratio).toBe(1);
  });

  it('dépassement : ratio plafonné à 1, done vrai', () => {
    const p = computeProgress(goal({ target: 3 }), 5);
    expect(p.ratio).toBe(1);
    expect(p.done).toBe(true);
  });

  it('cible nulle ou absente : ratio 0, jamais atteint (pas de division par zéro)', () => {
    const p = computeProgress(goal({ target: 0 }), 2);
    expect(p.ratio).toBe(0);
    expect(p.done).toBe(false);
  });
});

describe('describeGoal / formatGoalValue', () => {
  it('libellés selon métrique et période', () => {
    expect(describeGoal(goal({ metric: 'sessions', activity: 'velo', target: 3 }))).toBe(
      '3 sorties vélo / semaine',
    );
    expect(describeGoal(goal({ metric: 'sessions', activity: 'muscu', period: 'month', target: 8 }))).toBe(
      '8 séances muscu / mois',
    );
    expect(describeGoal(goal({ metric: 'distance', period: 'month', target: 100 }))).toBe(
      '100 km / mois',
    );
    expect(describeGoal(goal({ metric: 'tonnage', period: 'month', target: 5000 }))).toBe(
      '5000 kg soulevés / mois',
    );
  });

  it('valeur réalisée formatée avec unité', () => {
    expect(formatGoalValue(goal({ metric: 'sessions' }), 2)).toBe('2');
    expect(formatGoalValue(goal({ metric: 'distance' }), 42.5)).toBe('42.5 km');
    expect(formatGoalValue(goal({ metric: 'tonnage' }), 5123.4)).toBe('5123 kg');
  });
});

describe('parseGoals (robuste) / serializeGoals', () => {
  it('valeur absente ou JSON corrompu → liste vide', () => {
    expect(parseGoals(null)).toEqual([]);
    expect(parseGoals('not json')).toEqual([]);
    expect(parseGoals('{"not":"array"}')).toEqual([]);
  });

  it('ignore les entrées invalides et conserve les valides', () => {
    const raw = JSON.stringify([
      { id: 'a', metric: 'sessions', period: 'week', target: 3, activity: 'velo' }, // ok
      { id: 'b', metric: 'bogus', period: 'week', target: 3, activity: 'all' }, // métrique invalide
      { id: 'c', metric: 'distance', period: 'year', target: 50, activity: 'all' }, // période invalide
      { id: 'd', metric: 'tonnage', period: 'month', target: 0, activity: 'all' }, // cible <= 0
      { id: 'e', metric: 'distance', period: 'month', target: 100 }, // activity manquante → 'all'
    ]);
    const goals = parseGoals(raw);
    expect(goals.map((g) => g.id)).toEqual(['a', 'e']);
    expect(goals[1].activity).toBe('all');
  });

  it('round-trip serialize → parse', () => {
    const goals: Goal[] = [
      goal({ id: 'x', metric: 'distance', period: 'month', target: 120, activity: 'all' }),
    ];
    expect(parseGoals(serializeGoals(goals))).toEqual(goals);
  });
});

describe('upsertGoal / removeGoal / makeGoal', () => {
  it('ajoute un nouvel objectif', () => {
    const next = upsertGoal([], goal({ id: 'g1' }));
    expect(next).toHaveLength(1);
  });

  it('remplace l’objectif de même id', () => {
    const next = upsertGoal([goal({ id: 'g1', target: 3 })], goal({ id: 'g1', target: 5 }));
    expect(next).toHaveLength(1);
    expect(next[0].target).toBe(5);
  });

  it('supprime par id', () => {
    const next = removeGoal([goal({ id: 'g1' }), goal({ id: 'g2' })], 'g1');
    expect(next.map((g) => g.id)).toEqual(['g2']);
  });

  it('makeGoal attribue un id unique', () => {
    const a = makeGoal({ metric: 'sessions', period: 'week', target: 3, activity: 'all' });
    const b = makeGoal({ metric: 'sessions', period: 'week', target: 3, activity: 'all' });
    expect(a.id).not.toBe(b.id);
    expect(a.id).toBeTruthy();
  });
});

File: app/src/test/java/ovh/battistella/elan/domain/GoalsTest.kt
// Tests de la logique pure des objectifs d'entraînement (Goals.kt) : bornes de
// période, calcul d'avancement, libellés, parsing robuste et opérations de
// liste. Robolectric uniquement pour `org.json` (parseGoals / serializeGoals).
// Dates construites en heure locale (fuseau par défaut du système, comme
// `startOfWeekMs`) pour rester indépendant du fuseau de la CI.
package ovh.battistella.elan.domain

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId

@RunWith(RobolectricTestRunner::class)
class GoalsTest {
    private val zone: ZoneId = ZoneId.systemDefault()

    private fun goal(
        id: String = "g1",
        metric: GoalMetric = GoalMetric.SESSIONS,
        period: GoalPeriod = GoalPeriod.WEEK,
        target: Double = 3.0,
        activity: GoalActivity = GoalActivity.ALL,
    ) = Goal(id, metric, period, target, activity)

    private fun localMs(y: Int, m: Int, d: Int, h: Int = 0, min: Int = 0): Long =
        LocalDateTime.of(y, m, d, h, min).atZone(zone).toInstant().toEpochMilli()

    private fun local(ms: Long) = Instant.ofEpochMilli(ms).atZone(zone)

    // --- periodRange ---------------------------------------------------------

    @Test
    fun `semaine - lundi 00h00 vers lundi suivant (7 jours)`() {
        val now = localMs(2025, 6, 11, 14, 30) // mercredi 11 juin
        val (fromMs, toMs) = periodRange(GoalPeriod.WEEK, now)
        val from = local(fromMs)
        assertEquals(1, from.dayOfWeek.value) // lundi
        assertEquals(9, from.dayOfMonth)
        assertEquals(0, from.hour)
        assertEquals(7 * 86_400_000L, toMs - fromMs)
    }

    @Test
    fun `semaine - la borne de fin est re-calée sur le lundi suivant à minuit local`() {
        // Semaine du passage à l'heure d'été (30 mars 2025 en Europe) : avec un
        // +168 h fixe la borne dériverait d'une heure ; on exige lundi 00:00 local.
        val now = localMs(2025, 3, 26, 10, 0) // mercredi
        val (fromMs, toMs) = periodRange(GoalPeriod.WEEK, now)
        val from = local(fromMs)
        val to = local(toMs)
        assertEquals(1, to.dayOfWeek.value)
        assertEquals(0, to.hour)
        assertEquals(0, to.minute)
        assertEquals(from.toLocalDate().plusDays(7), to.toLocalDate())
        assertEquals(31, to.dayOfMonth)
    }

    @Test
    fun `mois - 1er du mois vers 1er du mois suivant`() {
        val now = localMs(2025, 6, 11, 14, 30) // juin
        val (fromMs, toMs) = periodRange(GoalPeriod.MONTH, now)
        val from = local(fromMs)
        val to = local(toMs)
        assertEquals(1, from.dayOfMonth)
        assertEquals(6, from.monthValue) // juin
        assertEquals(0, from.hour)
        assertEquals(1, to.dayOfMonth)
        assertEquals(7, to.monthValue) // juillet
    }

    @Test
    fun `mois - décembre déborde sur janvier de l'année suivante`() {
        val now = localMs(2025, 12, 20) // décembre 2025
        val (_, toMs) = periodRange(GoalPeriod.MONTH, now)
        val to = local(toMs)
        assertEquals(2026, to.year)
        assertEquals(1, to.monthValue) // janvier
    }

    // --- computeProgress -----------------------------------------------------

    @Test
    fun `avancement partiel - ratio borné, non atteint`() {
        val p = computeProgress(goal(target = 4.0), 1.0)
        assertEquals(0.25, p.ratio, 1e-9)
        assertFalse(p.done)
        assertEquals(1.0, p.value, 0.0)
        assertEquals(4.0, p.target, 0.0)
    }

    @Test
    fun `atteint à exactement la cible`() {
        val p = computeProgress(goal(target = 3.0), 3.0)
        assertTrue(p.done)
        assertEquals(1.0, p.ratio, 0.0)
    }

    @Test
    fun `dépassement - ratio plafonné à 1, done vrai`() {
        val p = computeProgress(goal(target = 3.0), 5.0)
        assertEquals(1.0, p.ratio, 0.0)
        assertTrue(p.done)
    }

    @Test
    fun `cible nulle - ratio 0, jamais atteint (pas de division par zéro)`() {
        val p = computeProgress(goal(target = 0.0), 2.0)
        assertEquals(0.0, p.ratio, 0.0)
        assertFalse(p.done)
    }

    // --- describeGoal / formatGoalValue --------------------------------------

    @Test
    fun `libellés selon métrique et période`() {
        assertEquals(
            "3 sorties vélo / semaine",
            describeGoal(goal(metric = GoalMetric.SESSIONS, activity = GoalActivity.VELO, target = 3.0)),
        )
        assertEquals(
            "8 séances muscu / mois",
            describeGoal(goal(metric = GoalMetric.SESSIONS, activity = GoalActivity.MUSCU, period = GoalPeriod.MONTH, target = 8.0)),
        )
        assertEquals("100 km / mois", describeGoal(goal(metric = GoalMetric.DISTANCE, period = GoalPeriod.MONTH, target = 100.0)))
        assertEquals(
            "5000 kg soulevés / mois",
            describeGoal(goal(metric = GoalMetric.TONNAGE, period = GoalPeriod.MONTH, target = 5000.0)),
        )
        assertEquals("3 séances / semaine", describeGoal(goal()))
    }

    @Test
    fun `valeur réalisée formatée avec unité`() {
        assertEquals("2", formatGoalValue(goal(metric = GoalMetric.SESSIONS), 2.0))
        assertEquals("42.5 km", formatGoalValue(goal(metric = GoalMetric.DISTANCE), 42.5))
        assertEquals("5123 kg", formatGoalValue(goal(metric = GoalMetric.TONNAGE), 5123.4))
    }

    // --- parseGoals (robuste) / serializeGoals -------------------------------

    @Test
    fun `valeur absente ou JSON corrompu - liste vide`() {
        assertEquals(emptyList<Goal>(), parseGoals(null))
        assertEquals(emptyList<Goal>(), parseGoals(""))
        assertEquals(emptyList<Goal>(), parseGoals("not json"))
        assertEquals(emptyList<Goal>(), parseGoals("""{"not":"array"}"""))
    }

    @Test
    fun `ignore les entrées invalides et conserve les valides`() {
        val raw = """[
            {"id":"a","metric":"sessions","period":"week","target":3,"activity":"velo"},
            {"id":"b","metric":"bogus","period":"week","target":3,"activity":"all"},
            {"id":"c","metric":"distance","period":"year","target":50,"activity":"all"},
            {"id":"d","metric":"tonnage","period":"month","target":0,"activity":"all"},
            {"id":"e","metric":"distance","period":"month","target":100}
        ]"""
        val goals = parseGoals(raw)
        assertEquals(listOf("a", "e"), goals.map { it.id })
        assertEquals(GoalActivity.ALL, goals[1].activity)
        assertEquals(100.0, goals[1].target, 0.0)
    }

    @Test
    fun `régénère un id absent et rejette une cible non numérique`() {
        val raw = """[
            {"metric":"sessions","period":"week","target":"3"},
            {"id":"","metric":"sessions","period":"week","target":2},
            {"id":"z","metric":"sessions","period":"week","target":"abc"}
        ]"""
        val goals = parseGoals(raw)
        assertEquals(2, goals.size)
        assertTrue(goals[0].id.startsWith("g"))
        assertTrue(goals[1].id.startsWith("g"))
        assertNotEquals(goals[0].id, goals[1].id)
        assertEquals(3.0, goals[0].target, 0.0)
    }

    @Test
    fun `round-trip serialize vers parse`() {
        val goals = listOf(
            goal(id = "x", metric = GoalMetric.DISTANCE, period = GoalPeriod.MONTH, target = 120.0, activity = GoalActivity.ALL),
            goal(id = "y", metric = GoalMetric.TONNAGE, period = GoalPeriod.WEEK, target = 2500.5, activity = GoalActivity.MUSCU),
        )
        assertEquals(goals, parseGoals(serializeGoals(goals)))
    }

    // --- upsertGoal / removeGoal / makeGoal ----------------------------------

    @Test
    fun `ajoute un nouvel objectif`() {
        assertEquals(1, upsertGoal(emptyList(), goal(id = "g1")).size)
    }

    @Test
    fun `remplace l'objectif de même id`() {
        val next = upsertGoal(listOf(goal(id = "g1", target = 3.0)), goal(id = "g1", target = 5.0))
        assertEquals(1, next.size)
        assertEquals(5.0, next[0].target, 0.0)
    }

    @Test
    fun `supprime par id`() {
        val next = removeGoal(listOf(goal(id = "g1"), goal(id = "g2")), "g1")
        assertEquals(listOf("g2"), next.map { it.id })
    }

    @Test
    fun `makeGoal attribue un id unique`() {
        val a = makeGoal(GoalMetric.SESSIONS, GoalPeriod.WEEK, 3.0, GoalActivity.ALL)
        val b = makeGoal(GoalMetric.SESSIONS, GoalPeriod.WEEK, 3.0, GoalActivity.ALL)
        assertNotEquals(a.id, b.id)
        assertTrue(a.id.isNotEmpty())
        assertTrue(a.id.startsWith("g"))
    }

    // --- libellés des nouvelles activités ------------------------------------

    @Test
    fun `nomme la course et la marche dans le libellé d'objectif`() {
        assertTrue(
            describeGoal(Goal("a", GoalMetric.SESSIONS, GoalPeriod.WEEK, 3.0, GoalActivity.COURSE))
                .contains("sorties course à pied"),
        )
        assertTrue(
            describeGoal(Goal("b", GoalMetric.SESSIONS, GoalPeriod.MONTH, 8.0, GoalActivity.MARCHE))
                .contains("sorties marche"),
        )
    }

    @Test
    fun `accepte course et marche à la relecture des réglages`() {
        val raw = """[
            {"id":"a","metric":"distance","period":"month","target":50,"activity":"course"},
            {"id":"b","metric":"sessions","period":"week","target":2,"activity":"marche"}
        ]"""
        assertEquals(2, parseGoals(raw).size)
        assertEquals(GoalActivity.COURSE, parseGoals(raw)[0].activity)
    }

    @Test
    fun `ramène une activité inconnue sur toutes sans perdre l'objectif`() {
        // Lecture défensive : un réglage restauré d'une autre version ne doit pas
        // disparaître silencieusement — il retombe sur le comptage le plus large.
        val raw = """[{"id":"c","metric":"sessions","period":"week","target":2,"activity":"natation"}]"""
        assertEquals(1, parseGoals(raw).size)
        assertEquals(GoalActivity.ALL, parseGoals(raw)[0].activity)
    }
}
